### PR TITLE
Let non-viewable Popup Annotations inherit the parent's Annotation Flags if the parent is viewable

### PR DESCRIPTION
**Pull request by Snuffleupagus**
Fixes http://www.pdf-archive.com/2013/09/30/file2/file2.pdf.

Note how it's not possible to show the various Popup Annotations in the above document.
To fix that, this patch lets the Popup inherit the flags of the parent, in the special case where the parent is `viewable` _and_ the Popup is not.
In general, I don't think that a Popup must have the same flags set as the parent. However, it seems very strange to have a `viewable` parent annotation, and then not being able to view the Popup.

Annoyingly the PDF specification doesn't, as far as I can find, mention anything about how this case should be handled, but this patch seem consistent with the actual behaviour in Adobe Reader.

@timvandermeij How do you feel about this patch? Should we consider fixing this edge-case, or are we more worried about breakage elsewhere.
~~_Please note:_ I've not attached a test-case, nor run the tests, since I wanted feedback on the general idea of the patch before spending a lot of time on this.~~

### CODE REVIEW

**Comment by timvandermeij** [comment]
I think the general idea here is fine. Okular is also able to properly view this file, so I think it makes sense to fix this, even though (luckily) it's very much an edge case. The situation where the parent is visible but the popup is not is sufficiently rare that I'm not worried about breakage in other areas.

File: src/core/annotation.js [hunk 1]
```diff
@@ -55,6 +55,22 @@ var ObjectLoader = coreObj.ObjectLoader;
 var FileSpec = coreObj.FileSpec;
 var OperatorList = coreEvaluator.OperatorList;
 
+function hasFlag(flags, bit) {
```
Comment (timvandermeij) [inline src/core/annotation.js]: Can't we keep this inside the class, but instead use them via private methods? So, `get viewable()` would use `return this._isViewable(this.flags)` and `_isViewable` is a private method of the `Annotation` class. It seems to make sense to me to keep this inside the `Annotation` class as that is essentially where it operates on. All other classes inherit from `Annotation`, so they will also have these private methods.

File: src/core/annotation.js [hunk 2]
```diff
@@ -224,11 +236,7 @@ var Annotation = (function AnnotationClosure() {
      * @see {@link shared/util.js}
      */
     setFlags: function Annotation_setFlags(flags) {
-      if (isInt(flags)) {
-        this.flags = flags;
-      } else {
-        this.flags = 0;
-      }
+      this.flags = (isInt(flags) && flags > 0) ? flags : 0;
```
Comment (timvandermeij) [inline src/core/annotation.js]: Would this also address #7280, at least partially?

File: src/core/annotation.js (hunk 1 above)
Comment (Snuffleupagus) [inline src/core/annotation.js]: > Can we keep these functions inside the class, but instead access them via private methods?

Of course we can, thanks for suggesting this!
I wasn't entirely happy about moving that code out of the `Annotation` class, so nice idea :-)

File: src/core/annotation.js (hunk 2 above)
Comment (Snuffleupagus) [inline src/core/annotation.js]: I suppose so, together with the slight change of the `_hasFlag` method in the new patch.
I should perhaps not have sneaked this into the current patch, but given that I needed to touch most of those methods anyway, it seemed better to reduce code churn a bit.

File: src/core/annotation.js [hunk 3]
```diff
@@ -191,27 +191,48 @@ var Annotation = (function AnnotationClosure() {
 
   Annotation.prototype = {
     /**
+     * @private
+     */
+    _hasFlag: function Annotation_isViewable(flags, flag) {
```
Comment (timvandermeij) [inline src/core/annotation.js]: This should be `Annotation_hasFlag`.

Comment (Snuffleupagus) [inline src/core/annotation.js]: Whoops, fixed now.

**Comment by timvandermeij** [comment]
Thank you for working on this!
